File: features/document/src/main/java/com/nivasafinance/features/document/exception/S3UrlGenerationException.java
package com.nivasafinance.features.document.exception;

import org.springframework.context.MessageSource;

import java.io.Serial;

public class S3UrlGenerationException extends S3OperationException {
    
    @Serial
    private static final long serialVersionUID = 1L;
    
    public static final String S3_URL_GENERATION_FAILED_KEY = "error.document.s3.url.generation.failed";
    
    public S3UrlGenerationException(String message) {
        super(message);
    }
    
    public S3UrlGenerationException(MessageSource messageSource) {
        super(S3_URL_GENERATION_FAILED_KEY, messageSource);
    }
    
    public S3UrlGenerationException(String error, MessageSource messageSource) {
        super(S3_URL_GENERATION_FAILED_KEY, error, messageSource);
    }
}

File: features/document/src/main/java/com/nivasafinance/features/document/exception/S3UrlGenerationException.kt
package com.nivasafinance.features.document.exception

import com.nivasafinance.common.exception.BadRequestException
import org.springframework.context.MessageSource
import org.springframework.context.i18n.LocaleContextHolder
import java.util.Locale

class S3UrlGenerationException(
    message: String
) : BadRequestException(message) {

    companion object {
        private val locale: Locale = LocaleContextHolder.getLocale()
        const val S3_URL_GENERATION_FAILED_KEY = "error.document.s3.url.generation.failed"
    }

    constructor(messageSource: MessageSource) : this(
        messageSource.getMessage(S3_URL_GENERATION_FAILED_KEY, arrayOf(""), locale)
    )

    constructor(error: String, messageSource: MessageSource) : this(
        messageSource.getMessage(S3_URL_GENERATION_FAILED_KEY, arrayOf(error), locale)
    )
}
